src/perform_step/adams_bashforth_moulton_perform_step.jl

```diff
@@ -1326,3 +1326,138 @@ end
     f(k4,u,p,t+dt)
   end
 end
+
+# VCABM
+
+function initialize!(integrator,cache::VCABMConstantCache)
+  integrator.fsalfirst = integrator.f(integrator.uprev, integrator.p, integrator.t) # Pre-start fsal
+  integrator.kshortsize = 2
+  integrator.k = typeof(integrator.k)(integrator.kshortsize)
+
+  # Avoid undefined entries if k is an array of arrays
+  integrator.fsallast = zero(integrator.fsalfirst)
+  integrator.k[1] = integrator.fsalfirst
+  integrator.k[2] = integrator.fsallast
+end
+
+@muladd function perform_step!(integrator,cache::VCABMConstantCache,repeat_step=false)
+  @unpack t,dt,uprev,u,f,p = integrator
+  @unpack dts,g,ϕ_n,ϕ_np1,ϕstar_n,ϕstar_nm1,order,max_order = cache
+  k1 = integrator.fsalfirst
+  step = integrator.iter
+  k = order
+  for i = 13:-1:2
+    dts[i] = dts[i-1]
+  end
+  dts[1] = dt
+  ϕ_and_ϕstar!(cache,k1,k)
+  g_coefs!(cache,k+1)
+  u = uprev
+  for i = 1:(k-1)
+      u += dt * g[i] * ϕstar_n[i]
+  end
+  du_np1 = f(u,p,t+dt)
+  ϕ_np1!(cache, du_np1, k+1)
+  u += dt * g[k] * ϕ_np1[k]
+  if integrator.opts.adaptive
+    utilde = dt * (g[k+1]-g[k]) * ϕ_np1[k+1]
+    atmp = calculate_residuals(utilde, uprev, u, integrator.opts.abstol, integrator.opts.reltol,integrator.opts.internalnorm)
+    integrator.EEst = integrator.opts.internalnorm(atmp)
+    if integrator.EEst >= one(integrator.EEst)
+      return nothing
+    end
+    integrator.fsallast = f(u, p, t+dt)
+    if step <= 4
+      cache.order = min(order+1,3)
+    else
+      # utildem2 = dt * γstar[(k-2)+1] * ϕ_np1[k-1]
+      utildem2 = dt * (g[k-1]-g[k-2]) * ϕ_np1[k-1]
+      # utildem1 = dt * γstar[(k-1)+1] * ϕ_np1[k]
+      utildem1 = dt * (g[k]-g[k-1]) * ϕ_np1[k]
+      ϕ_and_ϕstar!(cache,k1,k+1)
+      ϕ_np1!(cache, integrator.fsallast, k+2)
+      utildep1 = dt * γstar[(k+1)+1] * ϕ_np1[k+2]
+      atmpm2 = calculate_residuals(utildem2, uprev, u, integrator.opts.abstol, integrator.opts.reltol,integrator.opts.internalnorm)
+      atmpm1 = calculate_residuals(utildem1, uprev, u, integrator.opts.abstol, integrator.opts.reltol,integrator.opts.internalnorm)
+      atmpp1 = calculate_residuals(utildep1, uprev, u, integrator.opts.abstol, integrator.opts.reltol,integrator.opts.internalnorm)
+      errm2 = integrator.opts.internalnorm(atmpm2)
+      errm1 = integrator.opts.internalnorm(atmpm1)
+      errp1 = integrator.opts.internalnorm(atmpp1)
+      if max(errm2,errm1) <= integrator.EEst
+        cache.order = order - 1
+      elseif errp1 < integrator.EEst
+        cache.order = min(order+1,max_order)
```

Comment:
Hairer notes that when the order is changed the stepsize is kept constant. This is done because the higher order method has a smaller stability region, so changing both can cause it to go unstable. To hack this in, just do `integrator.EEst = one(integrator.EEst)` and add a comment saying that this is for keeping the stepsize constant in the next step.